src/mlpack/methods/ann/layer/batch_norm.hpp

```diff
@@ -68,15 +66,44 @@ class BatchNorm : public Layer<InputType, OutputType>
    *                updating the parameters or momentum is used.
    * @param momentum Parameter used to to update the running mean and variance.
    */
-  BatchNorm(const size_t size,
-            const double eps = 1e-8,
-            const bool average = true,
-            const double momentum = 0.1);
+  BatchNormType(const size_t size,
+                const double eps = 1e-8,
+                const bool average = true,
+                const double momentum = 0.1);
+  
+  virtual ~BatchNormType() { }
+
+  //! Clone the BatchNormType object. This handles polymorphism correctly.
+  BatchNormType* Clone() const { return new BatchNormType(*this); }
+
+  //! Copy the other BatchNorm layer (but not weights).
+  BatchNormType(const BatchNormType& layer);
+
+  //! Take ownership of the members of the other BatchNorm layer (but not weights).
+  BatchNormType(BatchNormType&& layer);
+
+  //! Copy the other BatchNorm layer (but not weights).
+  BatchNormType& operator=(const BatchNormType& layer);
+
+  //! Take ownership of the members of the other BatchNorm layer (but not weights).
+  BatchNormType& operator=(BatchNormType&& layer);
 
   /**
    * Reset the layer parameters.
    */
-  void SetWeights(typename OutputType::elem_type* weightsPtr);
+  void SetWeights(typename MatType::elem_type* weightsPtr);
+
+  /**
+   * Initialize the weight matrix of the layer.
+   *
+   * @param W Weight matrix to initialize.
+   * @param rows Number of rows.
+   * @param * (cols) Number of columns.
+   */
+  void CustomInitialize(
```

Comment:
I'm not sure I get the idea behind the custom weights initialization method, can you clarify.